Update Agent interface and improve documentation

cmcantalupo opened the pull request:
- Enables an Agent object to be used with a single controller to
  execute multiple run() calls.
- Fixes #1862 from github issues

cmcantalupo committed:
Update Agent interface and improve documentation

- Enables an Agent object to be used with a single controller to
  execute multiple run() calls.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix spelling issues in runtime documentation

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

fkeceli commented on service/geopmdpy/runtime.py:
```diff
@@ -265,9 +317,10 @@ def run(self):
             for loop_idx in TimedLoop(self._update_period, self._num_update):
                 if pid.poll() is not None:
-                    sys.stderr.write('Ths app process has ended.  return code = {}'.format(pid.returncode))
+                    sys.stderr.write('The app process has ended.  return code = {}'.format(pid.returncode))
```
Can you put a \n at the end of the string?

fkeceli reviewed (commented):
Few things:

- What about moving argv to run() method as I brought up before?
- In the Controller __init__ (...timeout) documentation it is mentioned that "the agent will run indefinitely (until killed by another process) if timeout == 0. I know that run will end if the app finishes naturally (not killed). I think that doc can be clearer.

cmcantalupo committed:
Update python controller test

- Add APIs to enable mutiple calls to Controller.run()

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add some logging to begin and end of run

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Read instructions retired for all package in the python agent example

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix test_controller issue just introduced

- Clean up print output a bit.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not throw on calls to ServiceIOGroup::read_batch()/write_batch()

- These methods are called even if no signals or controls
  are pushed.
- This will lead the controller to throw during regular
  operation if the ServiceIOGroup can be loaded.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Make argv a Controller.run() argument instead of a constructor arguement

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clarify documentation for Controller timeout parameter

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
Thanks for the feedback @fkeceli.  I think the last two patches address your comments.

fkeceli reviewed (dismissed):
Good changes. Thanks Chris. I actually hit "Do not throw on calls to ServiceIOGroup::read_batch()/write_batch()" bug in Saruman.Thanks for making the change.

bgeltz commented on service/geopmdpy/runtime.py:
```diff
@@ -181,7 +227,7 @@ def update(self, signals):
 
     def get_period(self):
-        """Get the target time interal for the control loop
+        """Get the target time internal for the control loop
```
```suggestion
        """Get the internal target time for the control loop
```
or just
```suggestion
        """Get the target time for the control loop
```

It's a getter.  It's assumed it's returning internal state.

cmcantalupo replied:
I think that was supposed to be *interval* not internal

bgeltz commented on service/integration/test/test_controller.py:
```diff
@@ -69,24 +83,29 @@ def get_report(self):
-    return err
+    controller = geopmdpy.runtime.Controller(agent)
+    print(controller.run(sys.argv[2:], cpu_frequency_max))
+    return controller.returncode()
```
Nice.

bgeltz commented on service/src/ServiceIOGroup.cpp:
```diff
@@ -171,14 +171,12 @@ namespace geopm
     {
-        throw Exception("ServiceIOGroup::read_batch()",
-                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
+
```
Maybe at least a comment here saying these are still unimplemented, and not intentionally empty?

cmcantalupo committed:
Address review comments

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
